### PR TITLE
Attempt to reconnect when there is a connection timeout

### PR DESCRIPTION
**Pull request by mmandrus**
**Context**
The input server list is only parsed once, the result being that the IPs are cached for the remainder of the life of the client. If the server list contains FQDNs, for example with a k8s deployment, the client will not be able to recover its connection to memcached if the server IP changes. 

**What's changed**
If there is a timeout error during any memcached operation, the client will attempt to reconnect by reinitializing its server list and recaching the server IPs. We perform this operation in the background so as not to delay cache responses. Consumers can also initialize the client with a context, which will start a cron job to periodically ping the server list and trigger a background reconnect if necessary.

### CODE REVIEW

File: memcache/memcache.go [hunk 1]
```diff
@@ -130,7 +131,30 @@ var (
...
+	return c
+}
+
+// NewWithBackgroundReconnect returns the same thing as New(server ...string), but takes a context to initiate a bakground reconnect cron
+func NewWithBackgroundReconnect(ctx context.Context, server ...string) *Client {
+	c := New(server...)
+
+	// periodically ping the provided servers -- if there are timeouts, it will trigger a reconnect attempt
+	go func() {
+		ticker := time.NewTicker(10 * time.Minute)
```
Comment (PoorlyDefinedBehaviour) [inline memcache/memcache.go:146]: I think `ticker.Stop()` needs to be called at some point

Comment (PoorlyDefinedBehaviour) [reply memcache/memcache.go:146]: Probably good to make it ticker interval configurable

File: memcache/memcache.go [hunk 2]
```diff
@@ -403,6 +430,9 @@ func (c *Client) dial(addr net.Addr) (net.Conn, error) {
 	}
 
 	if ne, ok := err.(net.Error); ok && ne.Timeout() {
+		// In the case of a timeout, we might need to reestablish a connection to one or more servers, for example in case of an IP change
+		// Fire this off in a goroutine so it doesn't delay a response back to the client
+		c.backgroundReconnect()
```
Comment (PoorlyDefinedBehaviour) [inline memcache/memcache.go:435]: It would be nice to test this

File: memcache/memcache.go (hunk 1 above)
Comment (mmandrus) [reply memcache/memcache.go]: Both done [428ba0a](https://github.com/grafana/gomemcache/pull/18/commits/428ba0a456d168730751c471bbede70b851031aa)

File: memcache/memcache.go (hunk 2 above)
Comment (mmandrus) [reply memcache/memcache.go:444]: Done [e764d08](https://github.com/grafana/gomemcache/pull/18/commits/e764d08ea2ede32d1d22550aaa14f81232b744a2)

**Review by PoorlyDefinedBehaviour — APPROVED** [review]

**Comment by 56quarters** [comment]
Just stumbled across this PR. FYI, the Memcached client wrapper in dskit already includes this functionality: https://github.com/grafana/dskit/blob/main/cache/memcached_client.go

**Comment by bboreham** [comment]
Next time try communicating with people who have previously made PRs to a repo.

**Pull request merged**